File: Hackerearth/Dynamic Programming/2 Dimensional/longest_increasing_path.py
# !/usr/bin/env python3
# -*- coding: utf-8 -*-


#	✪ H4WK3yE乡
#	Mohd. Farhan Tahir
#	Indian Institute Of Information Technology and Management,Gwalior


import sys

inf = float("inf")
mod = 1000000007


def get_array():
    return list(map(int, sys.stdin.readline().split()))


def input():
    return sys.stdin.readline()


def get_ints():
    return map(int, sys.stdin.readline().split())

# ///==========MAIN=============///


def solve():
    dp[1][1] = 1
    mx = 1
    for i in range(2, n+1):
        if matrix[i][1] > matrix[i-1][1]:
            dp[i][1] = 1+dp[i-1][1]
            mx = max(dp[i][1], mx)
        else:
            break
    for j in range(2, m+1):
        if matrix[1][j] > matrix[1][j-1]:
            dp[1][j] = 1+dp[1][j-1]
            mx = max(dp[1][j], mx)
        else:
            break
    '''for i in range(n+1):
        print(dp[i])
    print("\n")'''
    for i in range(2, n+1):
        for j in range(2, m+1):
            if matrix[i][j] > matrix[i-1][j] and dp[i-1][j] != 0:
                dp[i][j] = 1+dp[i-1][j]
            if matrix[i][j] > matrix[i][j-1] and dp[i][j-1] != 0:
                dp[i][j] = 1+dp[i][j-1]
            mx = max(dp[i][j], mx)

    '''for i in range(n+1):
        print(dp[i])'''
    return mx


if __name__ == "__main__":
    for tc in range(int(input())):
        n, m = get_ints()
        matrix = [[0 for _ in range(m+1)] for _ in range(n+1)]
        dp = [[0 for i in range(m+1)] for i in range(n+1)]
        for i in range(n):
            inp = get_array()
            for j in range(m):
                matrix[i+1][j+1] = inp[j]
        print(solve())

File: Hackerearth/DP 2_D/longest_increasing_path.cpp
/*
				C++
				encoding:UTF-8
				Modified: <10/May/2019 07:17:56 PM>

				✪ H4WK3yE乡
				Mohd. Farhan Tahir
				Indian Institute Of Information Technology (IIIT), Gwalior
*/

#include <bits/stdc++.h>

using namespace std;

// clang-format off

#define		 ll                     int64_t
#define		 mod                    1000000007
#define		 ve                     vector
#define		 pb                     push_back
#define		 umap                   unordered_map
#define		 pq                     priority_queue
#define		 endl                   "\n";
#define		 ff                     first
#define		 ss                     second
#define		 pii                    pair<int, int>
#define		 len(v)                 int(v.size())
#define		 all(v)                 v.begin(), v.end()
#define		 reset(a, b)            memset(a, b, sizeof(a));
#define		 sum(v)                 accumulate(v.begin(), v.end(), (ll)0);
#define		 pr1(x)                 cout<<x<<"\n"
#define		 pr2(x, y)              cout<<x<<" "<<y<<"\n"
#define		 pr3(x, y, z)           cout<<x<<" "<<y<<" "<<z<<"\n"
#define		 prv(v, s, n)           fr(i, s, n) cout<<v[i]<<" "; cout<<"\n"
#define		 fr(i, s, n)            for (int i = s ; i < n ; ++i)
#define		 dfr(i, s, n)           for (int i = s ; i > n ; --i)

ll powermod(ll _a,ll _b,ll _m) {ll _r=1;while(_b){if(_b%2==1)_r=(_r*_a)%_m;_b/=2;_a=(_a*_a)%_m;}return _r;}

// clang-format on

const int N = 105;
int matrix[N][N];
int n, m;
int dp[N][N];

int solve (int i, int j) {
	if (i == n + 1 or j == m + 1) return 0;
	if (matrix[i + 1][j] < matrix[i][j] and matrix[i][j + 1] < matrix[i][j]) return 0;
	if (dp[i][j] != 0) return dp[i][j];
	if (matrix[i + 1][j] > matrix[i][j]) dp[i][j] = max (dp[i][j], 1 + solve (i + 1, j));
	if (matrix[i][j + 1] > matrix[i][j]) dp[i][j] = max (dp[i][j], 1 + solve (i, j + 1));
	return dp[i][j];
}

int main () {
	ios_base::sync_with_stdio (false), cin.tie (0);
	int tc;
	cin >> tc;
	while (tc--) {
		reset (matrix, 0);
		reset (dp, 0);
		cin >> n >> m;
		fr (i, 1, n + 1) fr (j, 1, m + 1) cin >> matrix[i][j];
		pr1 (1 + solve (1, 1));
	}
	return 0;
}
